src/trg-client.c

```diff
@@ -480,360 +430,272 @@ void trg_client_configunlock(TrgClient *tc)
-
-            curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 0);
-            curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 0);
-        }
-
-        proxy = trg_client_get_proxy(tc);
-        if (proxy) {
-            curl_easy_setopt(curl, CURLOPT_PROXYTYPE, CURLPROXY_HTTP);
-            curl_easy_setopt(curl, CURLOPT_PROXY, proxy);
-        }
-
-        tls->serial = priv->configSerial;
-        priv->http_class = http_class;
-    }
-
-    if (http_class == HTTP_CLASS_TRANSMISSION)
-        curl_easy_setopt(curl, CURLOPT_URL, trg_client_get_url(tc));
-
-    curl_easy_setopt(curl, CURLOPT_TIMEOUT,
-                     (long)trg_prefs_get_int(prefs, TRG_PREFS_KEY_TIMEOUT, TRG_PREFS_CONNECTION));
-
-    curl_easy_setopt(curl, CURLOPT_ACCEPT_ENCODING, "gzip");
+    request->msg = NULL;
+    request->cancellable = NULL;
 
-    g_mutex_unlock(&priv->configMutex);
+    g_object_ref(tc);
+    request->client = tc;
 
-    /* Headers are set on each use, then freed, so make sure invalid headers aren't still around. */
-    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, NULL);
+    request->body = body;
+    request->response_cb = cb;
+    request->cb_data = cb_data;
 
-    return curl;
+    return request;
 }
 
-static int trg_http_perform_inner(TrgClient *tc, trg_request *request, trg_response *response,
-                                  gboolean recurse)
-{
-    CURL *curl = get_curl(tc, HTTP_CLASS_TRANSMISSION);
-    struct curl_slist *headers = NULL;
-    gchar *session_id = NULL;
-    long httpCode = 0;
-
-    response->size = 0;
-    response->raw = NULL;
+static void session_id_callback(SoupMessage *msg, gpointer user_data);
+static gboolean auth_callback(SoupMessage *msg, SoupAuth *auth, gboolean retry, gpointer user_data);
+static void rpc_callback(GObject *source, GAsyncResult *result, gpointer user_data);
+static gboolean tls_callback(SoupMessage *msg, GTlsCertificate *cert,
+                             GTlsCertificateFlags tls_errors, gpointer user_data);
 
-    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, request->body);
-    curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void *)response);
+static gboolean trg_request_setup_msg(trg_request *request)
+{
+    TrgClientPrivate *priv = request->client->priv;
+    SoupMessage *msg;
+    SoupMessageHeaders *req_headers;
 
-    session_id = trg_client_get_session_id(tc);
-    if (session_id) {
-        headers = curl_slist_append(NULL, session_id);
-        g_free(session_id);
+    /* setup_msg is also called if a previous message failed, so we need to clear out
+     * the previous one (I think) */
+    if (request->msg) {
+        g_clear_pointer(&request->msg, g_object_unref);
+        g_clear_pointer(&request->cancellable, g_object_unref);
     }
 
-    trg_client_inject_custom_headers(tc, &headers);
-
-    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
-
-    response->status = curl_easy_perform(curl);
+    msg = soup_message_new(SOUP_METHOD_POST, priv->url);
```

Comment:
I mentioned before but you can make `priv->url` a `GUri*` and change this to `soup_message_new_from_uri()` which cannot fail (Uri creation may fail but that happens earlier).

Reply:
I did consider using that, but it requires glib2 api 2.66+, which is currently not compatible with trg due to some depracated macro usage when making private methods/attributes in subclassing. It's on my TODO list to fix these classes and modernize the subclassing, so that we can drop the `-DGLIB_VERSION_MAX_ALLOWED` shenanigans as well and just rely on meson to enforce a minimum API version. 

For now, this will have to do.

Reply:
Oh thats not a problem. I pushed a commit bumping this.

Reply:
Don't drop `GLIB_VERSION_` macros btw. Some features are behind them and they are only helpful for warnings about using too new or old features.